lazy_dataset/core.py

```diff
@@ -1132,6 +1132,90 @@ def apply(self, apply_fn: callable) -> 'Dataset':
+        """
+        Caches data in memory. The dataset has to be indexable because the
+        cache needs a unique identifier (key) for each example.
+
+        Warnings:
+            This dataset is *not* immutable! It maintains the cache as an
+            instance variable.
+
+            This dataset freezes everything that comes before this dataset!
+            E.g., anything random before applying `.cache` is frozen. Any
+            shuffling has to be applied after caching!
+
+            `keep_mem_free` has no effect, if `lazy=False`! If `lazy=False`,
+            the dataset will always load all data.
+
+        Examples:
+
+            `Dataset.cache` can give a dataset that uses filter excpetions and
+            thus has an unkown effective length the correct length again:
+            >>> ds = new({'a': 1, 'b': 2, 'c': 3, 'd': 4})
+            >>> def m(x):
+            ...     if x % 2:
+            ...         raise FilterException()
+            ...     return x
+            >>> ds = ds.map(m)
+            >>> ds = ds.cache(lazy=False, catch_filter_exception=True)
+            >>> len(ds)
+            2
+
+            Generate lots of data and hope that it doesn't crash
+            >>> ds = new(dict(zip(list(range(10000)), list(range(10000)))))
+            >>> import numpy as np
+            >>> ds = ds.map(lambda x: np.random.randn(1000, 1000, 1000))
+            >>> ds = ds.cache(keep_mem_free=(5, 'GB'))
+            >>> for example in ds:
+            ...     pass # ...
+
+        Args:
+            lazy: If `True`, it caches "on the fly" and respects the setting
+                of `keep_mem_free`. If `False`, it immediatly, on invocation
+                of `cache`, loads all examples from the dataset into memory
+                without respecting `keep_mem_free`.
+            keep_mem_free: Tuple of value and unit (value, "unit"). Unit can be
+                "GB" for absolute memory and "fraction" for a percentage of
+                memory, e.g. `(5, "GB")` always keeps (at least) 5GB of memory
+                free and `(0.5, "fraction")` keeps 50% of the memory free. Keep
+                in mind that this only controlls the memory usage of the cache
+                and that other programs or parts of the program can also use
+                up memory!
+            catch_filter_exception: If `True`, `FilterException`s are catched
+                and the element that raised the exception while processing is
+                dropped. This can also be set to a specific type (or a list of
+                types) of exceptions to catch. This option is not compatible
+                with `lazy=True`. The resulting dataset always has a length (in
+                contrast to `prefetch` or `catch`).
+        """
+        if lazy:
+            assert not catch_filter_exception
+            return CacheDataset(self, keep_mem_free)
+        else:
+            assert not keep_mem_free
+            assert self.indexable
```

Comment:
Should we introduce a `ordered` flag?
Currently, is `indexable` the best option, but in a few cases it is `ordered` and not `indexable`.

Reply:
The actual requirement is that each example has a unique identifier. This can be provided by a key (indexable; some datasets could actually provide a unique key but are not indexable, e.g., catch) or the ordering. I think that we should try to support as many input datasets as possible, so an ordered flag sounds reasonable. But even if we introduce it, there are still datasets that in theory could be cached but we can't detect it (e.g., catch, filter).

Reply:
Currently, you use the example_id (e.g. key) as cache identifier. When you change this an index, you can also support the other datasets.